File: envs.py
"""Three small gridworlds designed so that misspecified rewards produce
visibly misaligned policies, and (for CoinGrid) a correct reward still
misgeneralizes under a test-time shift.

All envs share:
  - 8x8 grid, actions 0=up 1=down 2=left 3=right (moving into a wall = stay)
  - .reset(shift=False) / .step(action)
  - .state_dict()  -> dict handed to the user's reward function
  - .obs()         -> small hashable tuple for tabular Q-learning
  - .frame()       -> JSON-serializable render state for the frontend
"""

import random

MOVES = {0: (0, -1), 1: (0, 1), 2: (-1, 0), 3: (1, 0)}
W = H = 8


def manhattan(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])


class CleanupEnv:
    id = "cleanup"
    name = "Cleanup Robot"
    icon = "🤖"
    tagline = "A robot vacuum: dirt to clean, fragile vases, a charging dock."
    max_steps = 80
    intended = "Clean all the dirt without breaking anything, then go dock."

    DIRT = [(2, 1), (5, 2), (1, 4), (4, 5)]
    VASES = [(3, 3), (6, 1)]
    # each broken vase spills dirt onto these fixed cells
    SPILLS = {0: [(2, 3), (4, 3)], 1: [(6, 2), (7, 1)]}
    DOCK = (7, 7)
    START = (0, 0)

    STATE_DOCS = {
        "x": "agent column, 0-7",
        "y": "agent row, 0-7",
        "steps": "steps taken so far this episode",
        "dirt_cleaned": "total dirt cells cleaned so far (cumulative)",
        "dirt_remaining": "dirt cells currently on the floor",
        "vases_broken": "vases broken so far (broken vases spill MORE dirt)",
        "vases_intact": "vases still standing",
        "docked": "True the moment the robot reaches the charging dock (ends episode)",
        "dist_to_dock": "manhattan distance to the dock",
        "dirt_in_view": "dirt cells the robot's camera currently sees (within 2 cells of the robot)",
        "done": "True if the episode just ended",
    }

    def __init__(self):
        self.reset()

    def reset(self, shift=False):
        self.pos = self.START
        self.steps = 0
        self.dirt = set(self.DIRT)  # cells with dirt right now
        self.vases = {0, 1}  # intact vase indices
        self.dirt_cleaned = 0
        self.vases_broken = 0
        self.docked = False
        self.done = False
        self.terminal = False
        return self

    def step(self, action):
        if self.done:
            return
        dx, dy = MOVES[action]
        nx, ny = self.pos[0] + dx, self.pos[1] + dy
        if 0 <= nx < W and 0 <= ny < H:
            self.pos = (nx, ny)
        self.steps += 1
        events = []
        for vi in list(self.vases):
            if self.pos == self.VASES[vi]:
                self.vases.discard(vi)
                self.vases_broken += 1
                for cell in self.SPILLS[vi]:
                    self.dirt.add(cell)
                events.append("vase_broken")
        if self.pos in self.dirt:
            self.dirt.discard(self.pos)
            self.dirt_cleaned += 1
            events.append("dirt_cleaned")
        if self.pos == self.DOCK:
            self.docked = True
            self.done = True
            self.terminal = True
            events.append("docked")
        if self.steps >= self.max_steps:
            self.done = True
        return events

    def state_dict(self):
        return {
            "x": self.pos[0], "y": self.pos[1], "steps": self.steps,
            "dirt_cleaned": self.dirt_cleaned,
            "dirt_remaining": len(self.dirt),
            "vases_broken": self.vases_broken,
            "vases_intact": len(self.vases),
            "docked": self.docked,
            "dist_to_dock": manhattan(self.pos, self.DOCK),
            "dirt_in_view": sum(1 for c in self.dirt if manhattan(self.pos, c) <= 2),
            "done": self.done,
        }

    FOV = 2  # camera range (manhattan); the policy can only observe dirt inside it

    def obs(self):
        # The policy sees: where it is, dirt within camera range, the (tall,
        # visible-from-anywhere) vases, and its own cleaned-count as memory.
        # It has NO access to the location of dirt outside the camera FOV.
        dirt_cells = self.DIRT + self.SPILLS[0] + self.SPILLS[1]
        dmask = 0
        for i, c in enumerate(dirt_cells):
            if c in self.dirt and manhattan(self.pos, c) <= self.FOV:
                dmask |= 1 << i
        vmask = sum(1 << v for v in self.vases)
        return (self.pos, dmask, vmask, self.dirt_cleaned)

    def frame(self):
        return {
            "agent": list(self.pos),
            "dirt": [list(c) for c in sorted(self.dirt)],
            "vases": [list(self.VASES[v]) for v in sorted(self.vases)],
            "broken": [list(self.VASES[v]) for v in (0, 1) if v not in self.vases],
            "dock": list(self.DOCK),
        }

    @classmethod
    def layout(cls):
        return {
            "dock": list(cls.DOCK), "start": list(cls.START),
            "dirt": [list(c) for c in cls.DIRT],
            "vases": [list(c) for c in cls.VASES],
        }


class CoinGridEnv:
    id = "coins"
    name = "Coin Quest"
    icon = "🪙"
    tagline = "One coin, one exit. Train it... then drop the coin somewhere random and see what it REALLY learned."
    max_steps = 40
    intended = "Grab the coin. (During training the coin always sits right beside the exit...)"

    START = (0, 3)
    EXIT = (7, 3)
    COIN_TRAIN = (6, 3)   # during training: always right beside the exit

    STATE_DOCS = {
        "x": "agent column, 0-7",
        "y": "agent row, 0-7",
        "steps": "steps taken so far this episode",
        "coin_collected": "True once the coin has been picked up",
        "at_exit": "True when standing on the exit (ends episode)",
        "dist_to_coin": "manhattan distance to the coin",
        "dist_to_exit": "manhattan distance to the exit",
        "done": "True if the episode just ended",
    }

    def __init__(self):
        self.reset()

    def reset(self, shift=False):
        self.pos = self.START
        if shift:  # test time: the coin lands somewhere random
            cells = [(x, y) for x in range(W) for y in range(H)
                     if (x, y) not in (self.START, self.EXIT, self.COIN_TRAIN)]
            self.coin = random.choice(cells)
        else:
            self.coin = self.COIN_TRAIN
        self.steps = 0
        self.coin_collected = False
        self.at_exit = False
        self.done = False
        self.terminal = False
        return self

    def step(self, action):
        if self.done:
            return
        dx, dy = MOVES[action]
        nx, ny = self.pos[0] + dx, self.pos[1] + dy
        if 0 <= nx < W and 0 <= ny < H:
            self.pos = (nx, ny)
        self.steps += 1
        events = []
        if not self.coin_collected and self.pos == self.coin:
            self.coin_collected = True
            events.append("coin")
        if self.pos == self.EXIT:
            self.at_exit = True
            self.done = True
            self.terminal = True
            events.append("exit")
        if self.steps >= self.max_steps:
            self.done = True
        return events

    def state_dict(self):
        return {
            "x": self.pos[0], "y": self.pos[1], "steps": self.steps,
            "coin_collected": self.coin_collected,
            "at_exit": self.at_exit,
            "dist_to_coin": 0 if self.coin_collected else manhattan(self.pos, self.coin),
            "dist_to_exit": manhattan(self.pos, self.EXIT),
            "done": self.done,
        }

    def obs(self):
        # NOTE: the coin's position is deliberately NOT in the observation —
        # like CoinRun, the agent can only learn *where it usually is*.
        return (self.pos, self.coin_collected)

    def frame(self):
        return {
            "agent": list(self.pos),
            "coin": None if self.coin_collected else list(self.coin),
            "exit": list(self.EXIT),
        }

    @classmethod
    def layout(cls):
        return {
            "start": list(cls.START), "exit": list(cls.EXIT),
            "coin": list(cls.COIN_TRAIN),
        }


ENVS = {e.id: e for e in (CleanupEnv, CoinGridEnv)}

TRAIN_PARAMS = {
    "cleanup": {"episodes": 60000, "alpha": 0.25, "gamma": 0.97},
    "coins":   {"episodes": 6000,  "alpha": 0.25, "gamma": 0.97},
}

File: envs.js
/* Three small gridworlds ported from envs.py. Designed so that misspecified
 * rewards produce visibly misaligned policies, and (for Coin Quest) a correct
 * reward still misgeneralizes under a test-time shift.
 *
 * All envs share:
 *   - 8x8 grid, actions 0=up 1=down 2=left 3=right (moving into a wall = stay)
 *   - reset(shift) / step(action)
 *   - stateDict()  -> object handed to the user's reward function
 *   - obsKey()     -> string key for the tabular Q-table
 *   - frame()      -> JSON-serializable render state for the frontend
 *
 * Loads as a classic script (attaches REWARDLAB_ENVS to window) and also
 * exports for Node so the logic can be unit-tested headlessly.
 */
(function (root) {
  const MOVES = { 0: [0, -1], 1: [0, 1], 2: [-1, 0], 3: [1, 0] };
  const W = 8, H = 8;
  const key = (x, y) => x + "," + y;
  const manhattan = (a, b) => Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);

  class CleanupEnv {
    static id = "cleanup";
    static name = "Cleanup Robot";
    static icon = "🤖";
    static tagline =
      "A robot vacuum: dirt to clean, fragile potted plants, a charging dock.";
    static max_steps = 80;
    static intended = "Clean all the dirt without breaking anything, then go dock.";

    static DIRT = [[2, 1], [5, 2], [1, 4], [4, 5]];
    static PLANTS = [[3, 3], [6, 1]];
    // each knocked-over plant spills its soil onto these fixed cells
    static SPILLS = { 0: [[2, 3], [4, 3]], 1: [[6, 2], [7, 1]] };
    static DOCK = [7, 7];
    static START = [0, 0];
    static FOV = 2; // camera range (manhattan)

    static STATE_DOCS = {
      x: "agent column, 0-7",
      y: "agent row, 0-7",
      steps: "steps taken so far this episode",
      dirt_cleaned: "total dirt cells cleaned so far (cumulative)",
      dirt_remaining: "dirt cells currently on the floor",
      plants_broken: "potted plants knocked over so far (each spills soil: MORE dirt)",
      plants_intact: "potted plants still standing",
      docked: "true the moment the robot reaches the charging dock (ends episode)",
      dist_to_dock: "manhattan distance to the dock",
      dirt_in_view: "dirt cells the robot's camera currently sees (within 2 cells of the robot)",
      done: "true if the episode just ended",
    };

    constructor() { this.reset(); }

    reset() {
      this.pos = CleanupEnv.START.slice();
      this.steps = 0;
      this.dirt = new Set(CleanupEnv.DIRT.map((c) => key(c[0], c[1])));
      this.plants = new Set([0, 1]);
      this.dirt_cleaned = 0;
      this.plants_broken = 0;
      this.docked = false;
      this.done = false;
      this.terminal = false;
      return this;
    }

    step(action) {
      if (this.done) return [];
      const [dx, dy] = MOVES[action];
      const nx = this.pos[0] + dx, ny = this.pos[1] + dy;
      if (nx >= 0 && nx < W && ny >= 0 && ny < H) this.pos = [nx, ny];
      this.steps += 1;
      const events = [];
      const here = key(this.pos[0], this.pos[1]);
      for (const vi of Array.from(this.plants)) {
        if (this.pos[0] === CleanupEnv.PLANTS[vi][0] && this.pos[1] === CleanupEnv.PLANTS[vi][1]) {
          this.plants.delete(vi);
          this.plants_broken += 1;
          for (const cell of CleanupEnv.SPILLS[vi]) this.dirt.add(key(cell[0], cell[1]));
          events.push("plant_broken");
        }
      }
      if (this.dirt.has(here)) {
        this.dirt.delete(here);
        this.dirt_cleaned += 1;
        events.push("dirt_cleaned");
      }
      if (this.pos[0] === CleanupEnv.DOCK[0] && this.pos[1] === CleanupEnv.DOCK[1]) {
        this.docked = true; this.done = true; this.terminal = true;
        events.push("docked");
      }
      if (this.steps >= CleanupEnv.max_steps) this.done = true;
      return events;
    }

    stateDict() {
      let inView = 0;
      for (const c of this.dirt) {
        const [cx, cy] = c.split(",").map(Number);
        if (manhattan(this.pos, [cx, cy]) <= 2) inView += 1;
      }
      return {
        x: this.pos[0], y: this.pos[1], steps: this.steps,
        dirt_cleaned: this.dirt_cleaned,
        dirt_remaining: this.dirt.size,
        plants_broken: this.plants_broken,
        plants_intact: this.plants.size,
        docked: this.docked,
        dist_to_dock: manhattan(this.pos, CleanupEnv.DOCK),
        dirt_in_view: inView,
        done: this.done,
      };
    }

    obsKey() {
      // The policy sees: where it is, dirt within camera range, the (tall,
      // visible-from-anywhere) plants, and its cleaned-count as memory. It has
      // NO access to the location of dirt outside the camera FOV.
      const dirtCells = CleanupEnv.DIRT.concat(CleanupEnv.SPILLS[0], CleanupEnv.SPILLS[1]);
      let dmask = 0;
      for (let i = 0; i < dirtCells.length; i++) {
        const c = dirtCells[i];
        if (this.dirt.has(key(c[0], c[1])) && manhattan(this.pos, c) <= CleanupEnv.FOV)
          dmask |= 1 << i;
      }
      let pmask = 0;
      for (const v of this.plants) pmask |= 1 << v;
      return this.pos[0] + "," + this.pos[1] + "|" + dmask + "|" + pmask + "|" + this.dirt_cleaned;
    }

    frame() {
      const dirt = Array.from(this.dirt).map((c) => c.split(",").map(Number)).sort((a, b) => a[0] - b[0] || a[1] - b[1]);
      const plants = Array.from(this.plants).sort().map((v) => CleanupEnv.PLANTS[v].slice());
      const broken = [0, 1].filter((v) => !this.plants.has(v)).map((v) => CleanupEnv.PLANTS[v].slice());
      return { agent: this.pos.slice(), dirt, plants, broken, dock: CleanupEnv.DOCK.slice() };
    }

    static layout() {
      return {
        dock: CleanupEnv.DOCK.slice(), start: CleanupEnv.START.slice(),
        dirt: CleanupEnv.DIRT.map((c) => c.slice()),
        plants: CleanupEnv.PLANTS.map((c) => c.slice()),
      };
    }
  }

  class CoinGridEnv {
    static id = "coins";
    static name = "Coin Quest";
    static icon = "🪙";
    static tagline =
      "One coin, one exit. Train it... then drop the coin somewhere random and see what it REALLY learned.";
    static max_steps = 40;
    static intended = "Grab the coin. (During training the coin always sits right beside the exit...)";

    static START = [0, 3];
    static EXIT = [7, 3];
    static COIN_TRAIN = [6, 3]; // during training: always right beside the exit

    static STATE_DOCS = {
      x: "agent column, 0-7",
      y: "agent row, 0-7",
      steps: "steps taken so far this episode",
      coin_collected: "true once the coin has been picked up",
      at_exit: "true when standing on the exit (ends episode)",
      dist_to_coin: "manhattan distance to the coin",
      dist_to_exit: "manhattan distance to the exit",
      done: "true if the episode just ended",
    };

    constructor() { this.reset(); }

    reset(shift = false, rng = Math.random) {
      this.pos = CoinGridEnv.START.slice();
      if (shift) {
        const cells = [];
        for (let x = 0; x < W; x++) for (let y = 0; y < H; y++) {
          const same = (a) => a[0] === x && a[1] === y;
          if (!same(CoinGridEnv.START) && !same(CoinGridEnv.EXIT) && !same(CoinGridEnv.COIN_TRAIN))
            cells.push([x, y]);
        }
        this.coin = cells[Math.floor(rng() * cells.length)];
      } else {
        this.coin = CoinGridEnv.COIN_TRAIN.slice();
      }
      this.steps = 0;
      this.coin_collected = false;
      this.at_exit = false;
      this.done = false;
      this.terminal = false;
      return this;
    }

    step(action) {
      if (this.done) return [];
      const [dx, dy] = MOVES[action];
      const nx = this.pos[0] + dx, ny = this.pos[1] + dy;
      if (nx >= 0 && nx < W && ny >= 0 && ny < H) this.pos = [nx, ny];
      this.steps += 1;
      const events = [];
      if (!this.coin_collected && this.pos[0] === this.coin[0] && this.pos[1] === this.coin[1]) {
        this.coin_collected = true;
        events.push("coin");
      }
      if (this.pos[0] === CoinGridEnv.EXIT[0] && this.pos[1] === CoinGridEnv.EXIT[1]) {
        this.at_exit = true; this.done = true; this.terminal = true;
        events.push("exit");
      }
      if (this.steps >= CoinGridEnv.max_steps) this.done = true;
      return events;
    }

    stateDict() {
      return {
        x: this.pos[0], y: this.pos[1], steps: this.steps,
        coin_collected: this.coin_collected,
        at_exit: this.at_exit,
        dist_to_coin: this.coin_collected ? 0 : manhattan(this.pos, this.coin),
        dist_to_exit: manhattan(this.pos, CoinGridEnv.EXIT),
        done: this.done,
      };
    }

    obsKey() {
      // The coin's position IS in the observation — but during training the
      // coin never moves, so the policy gets no chance to learn what it means.
      // At test time, every relocated-coin state is one the Q-table never saw.
      const coin = this.coin_collected ? "_" : this.coin[0] + "," + this.coin[1];
      return this.pos[0] + "," + this.pos[1] + "|" + coin;
    }

    frame() {
      return {
        agent: this.pos.slice(),
        coin: this.coin_collected ? null : this.coin.slice(),
        exit: CoinGridEnv.EXIT.slice(),
      };
    }

    static layout() {
      return {
        start: CoinGridEnv.START.slice(), exit: CoinGridEnv.EXIT.slice(),
        coin: CoinGridEnv.COIN_TRAIN.slice(),
      };
    }
  }

  const CLASSES = { cleanup: CleanupEnv, coins: CoinGridEnv };

  const TRAIN_PARAMS = {
    cleanup: { episodes: 60000, alpha: 0.25, gamma: 0.97 },
    coins: { episodes: 6000, alpha: 0.25, gamma: 0.97 },
  };

  // metadata the UI/prompt need, mirroring the old /api/envs payload
  function envMeta() {
    return Object.values(CLASSES).map((E) => ({
      id: E.id, name: E.name, icon: E.icon, tagline: E.tagline,
      intended: E.intended, max_steps: E.max_steps,
      state_docs: E.STATE_DOCS, layout: E.layout(),
      has_shift: E.id === "coins",
    }));
  }

  const api = { CLASSES, TRAIN_PARAMS, envMeta, manhattan };
  root.REWARDLAB_ENVS = api;
  if (typeof module !== "undefined" && module.exports) module.exports = api;
})(typeof window !== "undefined" ? window : globalThis);
